### PR TITLE
add beforeFirstTestStatement, afterLastTestStatement to TestListener

### PR DESCRIPTION
**Pull request by MykolaGolubyev**
This is a prerequisite to #488 so we can fail as part of a test using a `TestListener`

### CODE REVIEW

File: webtau-core/src/main/java/com/twosigma/webtau/reporter/TestListener.java [hunk 1]
```diff
@@ -5,11 +5,40 @@
...
-    void beforeFirstTest();
+    /**
+     * before any test is ran
+     */
+    default void beforeFirstTest() {}
 
-    void beforeTestRun(WebTauTest test);
+    /**
+     * before test code is invoked
+     * @param test test
```
Comment (tsiq-clemens) [inline webtau-core/src/main/java/com/twosigma/webtau/reporter/TestListener.java]: these param lines don't add much to the code -- can we remove them or is some nit-picky style checker going to complain?

File: webtau-junit5/src/test/java/com/twosigma/webtau/junit5/WebTauJunitExtensionTest.groovy [hunk 2]
```diff
@@ -15,13 +18,47 @@ class WebTauJunitExtensionTest {
...
-        def report = JavaReport.INSTANCE.create()
 
-        def tests = report.tests.stream().collect(Collectors.toList())
-        tests.should == ['*id'                    | 'failed' | 'errored'] {
-                         ________________________________________________
-                         'beforeAllfailingInit'   | false    | true
-                         'afterAllfailingCleanup' | false    | true  }
+        reportTests.should == ['*id'                    | 'failed' | 'errored'] {
+                               ________________________________________________
+                               'beforeAllfailingInit'   | false    | true
```
Comment (tsiq-clemens) [inline webtau-junit5/src/test/java/com/twosigma/webtau/junit5/WebTauJunitExtensionTest.groovy]: nit: capitalize "Failing"

**Comment by tsiq-clemens** [comment]
This might be a good time to write a page on how to write/embed your own custom TestListener.
If the "owner" check (different PR) can be effectively conducted using one, I'd be happy to document the how-to.

File: webtau-core/src/main/java/com/twosigma/webtau/reporter/TestListener.java (hunk 1 above)
Comment (MykolaGolubyev) [reply webtau-core/src/main/java/com/twosigma/webtau/reporter/TestListener.java]: I believe javadoc will fail

File: webtau-junit5/src/test/java/com/twosigma/webtau/junit5/WebTauJunitExtensionTest.groovy (hunk 2 above)
Comment (MykolaGolubyev) [reply webtau-junit5/src/test/java/com/twosigma/webtau/junit5/WebTauJunitExtensionTest.groovy]: That’s just the way we generate id. Not supposed to be human readable.

File: webtau-core/src/main/java/com/twosigma/webtau/reporter/TestListener.java (hunk 1 above)
Comment (tsiq-karold) [reply webtau-core/src/main/java/com/twosigma/webtau/reporter/TestListener.java]: Javadoc will fail if you only do part of the javadoc.  If you don't have the comment at all, everything will be awesome.

**Comment by MykolaGolubyev** [comment]
PR to be able to pass TestListener through groovy config is coming up. Right now the only way is service loaders and it won’t be as convenient for groovy runner based tests.

**Pull request merged**